Add product form buttons and specific product edit page

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

https://user-images.githubusercontent.com/2240960/183488212-4146ab6e-4f0f-4d8a-bc72-9cffe68d86d1.mp4

The primary addition of this PR is a new edit screen for the new product page and the action buttons for the new form.

This did require numerous additional changes, one was the addition of the `Form` component and some changes in the data and component packages.

#34029  .

### How to test the changes in this Pull Request:

1. Load this branch and enable the `new-product-management-experience` feature flag by using the WCA Test Helper (Tools > WCA Test Helper > Features) and refresh the page.
2. Enable track logging by running this in your console: `localStorage.setItem( 'debug', 'wc-admin:*' );`
3. Navigate to **Products > Add New (MVP)** and type something in the name field
4. Click **Save draft** and a `wcadmin_product_edit` track should fire with some of the product details ( if a brand new product then these product details will be `undefined`).
5. Notice how a notice popped up with **Product successfully created** and the `path` in the url also changed to the new product id.
6. Edit the name again and click **Save draft** again, it should trigger a track again, this time with the product details defined.
7. Click **Preview** it should open a new tab with the product preview and the original tab should output a `wcadmin_product_preview_changes` track.
8. Click **Publish**, which should output a `wcadmin_product_update` track and also show a notice again **Product published.** with a **View in store** link that should open a new tab with a preview and also trigger the `wcadmin_product_preview_changes` track.
After publish the **Publish** button should also change to **Update**, the **Save draft** button should switch to **Switch to draft**
9. Click the dropdown beside the Update/Publish button and it should contain 3 options. Try each, they each should output their own tracks: `wcadmin_product_publish_and_copy`, `wcadmin_product_copy`, and `wcadmin_product_delete`
The two copy actions should create a new product with the `- Copy` tagged on to the name and redirected to it's edit screen (in draft).
10. When on an edit screen of a product click **Add New (MVP)** again, it should reset the form to pristine conditions again.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add initial product form action buttons to form

louwie17 committed:
Add new edit page and allow for updating and creating products

louwie17 committed:
Move crud functions to helper hook to keep things seperated better

louwie17 committed:
Add changelog

louwie17 committed:
ADd package changelogs

louwie17 committed:
Add isPending selector to products store

louwie17 committed:
Fix trash screen showing up when deleting product and add loading indicators

louwie17 committed:
Add extra types to Product type

louwie17 committed:
Update track names and add product data to tracks

louwie17 committed:
Update track logic when product is published

louwie17 committed:
Remove the Image section for now, to prevent confusion as it is not ready yet

louwie17 commented on packages/js/components/src/form/form.tsx:
```diff
@@ -61,6 +61,12 @@ type FormProps< Values > = {
+	value: T | ChangeEvent< HTMLInputElement >
+): value is ChangeEvent< HTMLInputElement > {
+	return ( value as ChangeEvent< HTMLInputElement > ).target !== undefined;
+}
```
Needed this as the `onChange` callback expected either a `string` or a `ChangeEvent` object.
Previously I was only expecting a `ChangeEvent` object, which didn't work when adding the Form to the product page.

louwie17 commented on packages/js/components/src/form/form.tsx:
```diff
@@ -96,6 +114,23 @@ function Form< Values extends Record< string, any > >(
+					...changedFields,
+					[ name ]: false,
+				} );
+			}
```
Added `changedFields` logic so we can determine if the form is `dirty` (has been changed) or not.

louwie17 commented on packages/js/data/src/products/action-types.ts:
```diff
@@ -1,4 +1,5 @@
 export enum TYPES {
+	CREATE_PRODUCT_START = 'CREATE_PRODUCT_START',
```
I added these events for any of the actions, as there was no way to tell when a product was being updated, created, or deleted. The `isResolving` or `hasFinishedResolution` only work for resolvers/selectors I guess.

In the reducer these events get toggled within the `pending` object (see the `reducer.ts` changes for this) and set back to `false` when the success or error actions are triggered.

louwie17 commented on packages/js/data/src/products/types.ts:
```diff
@@ -37,32 +43,65 @@ export type Product< Status = ProductStatus, Type = ProductType > = Omit<
+	'rating_count',
+	'related_ids',
+	'variations',
+] as const;
```
Moved these to a array, so I can use them more easily to remove these properties when updating or copying a product.

louwie17 commented on plugins/woocommerce-admin/client/products/edit-product-page.tsx:
```diff
@@ -0,0 +1,106 @@
+/**
```
The logic within this component could be combined and moved into the `add-product-page` component if we wanted to, but to keep things relatively clear I kept them separate for now.

louwie17 commented on plugins/woocommerce-admin/client/products/use-product-helper.ts:
```diff
@@ -0,0 +1,213 @@
+		: [];
+}
+
+export function useProductHelper() {
```
This is a custom hook I created to move some of the specific create, copy, update, delete logic so the `ProductFormActions` component doesn't get super long.
This would also be helpful in the case we need to call these functions from any other component (for example when we create the new product list page at some point).

louwie17 commented on plugins/woocommerce-admin/client/tasks/fills/experimental-products/card-layout.tsx:
```diff
@@ -41,7 +41,7 @@ const CardLayout: React.FC< CardProps > = ( { items } ) => {
 									} );
 									recordCompletionTime();
-									window.location = getAdminLink(
+									window.location.href = getAdminLink(
```
I updated the [/typings/index.d.ts](https://github.com/woocommerce/woocommerce/pull/34211/files#diff-4642695243afdb91bce1cf4782a7428a1e31e6c5e6d20f73c8196a169dbdaf49) file with the correct `window.location` type, which is why I updated a bunch of places to use `href` instead.

octaedro replied:
Do we also need to add the `href` [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/wp-admin-scripts/onboarding-homepage-notice/index.js#L85) and [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/tasks/fills/products/product-template-modal.js#L91)?

louwie17 replied:
Yeah probably although I do believe it will work the way it is (weirdly enough). I missed these as they were not `TS` files so the TypeScript didn't complain, but will update those. Thanks for catching that.

louwie17 replied:
Done as part of: https://github.com/woocommerce/woocommerce/pull/34211/commits/dedd1a6e021c8d0f47b3bd68a3c4fb99a8f82d54

louwie17 committed:
Add tests for the product form actions

louwie17 committed:
Update copy for publish & duplicate

octaedro commented on plugins/woocommerce/src/Admin/Features/Navigation/CoreMenu.php:
```diff
@@ -213,7 +213,8 @@ public static function get_items() {
 
-		$add_product_mvp = array();
+		$add_product_mvp  = array();
+		$edit_product_mvp = array();
```
I'm not sure whether we need to add this menu item. This would duplicate the `Add New (MVP)` item if it were added to the menu.

louwie17 replied:
Hmm good catch, for some reason I thought I need this and really thought I added the `$edit_product_mvp` item as part of the list below. But turns out I never did...
This should be good to remove then as it isn't doing anything.

louwie17 replied:
Done as part of: https://github.com/woocommerce/woocommerce/pull/34211/commits/3986d9db8db8ee536261f7fa78302474058ad6e8

octaedro reviewed (commented):
You did a great job here @louwie17! I just left a couple of minor comments; other than that, it's testing well here and code looks good!

louwie17 committed:
Remove unused code

louwie17 committed:
Set window.location correctly with href

louwie17 commented:
> You did a great job here @louwie17! I just left a couple of minor comments; other than that, it's testing well here and code looks good!

Thanks for the review @octaedro, I just addressed your feedback, this should be good for another look :)

octaedro reviewed (dismissed):
Good job @louwie17! LGTM 🚀

louwie17 commented:
@octaedro could I get a re-approval, I just rebased it as some of the checks were failing and wanted to make sure it wasn't related to my PR.

louwie17 committed:
Reset changes in pnpm lock

pmcpinto commented:
> I also made the Save draft button disappear once the product is published as it seemed strange to keep it (as an FYI, but would need a check by @jarekmorawski & @pmcpinto )

👍 makes sense cc @jarekmorawski

jarekmorawski commented:
> I also made the Save draft button disappear once the product is published as it seemed strange to keep it (as an FYI, but would need a check by @jarekmorawski & @pmcpinto )

Good call! The behavior of these buttons is supposed to mirror Gutenberg, where the `Save draft` button turns into `✓ Saved` when clicked, then becomes `Save draft` again when the user makes any changes. Can we replicate that here?

https://user-images.githubusercontent.com/79307566/183938488-340a6899-7789-4dd0-89e9-745be8945d2e.mov

louwie17 commented:
> Good call! The behavior of these buttons is supposed to mirror Gutenberg, where the `Save draft` button turns into `✓ Saved` when clicked, then becomes `Save draft` again when the user makes any changes. Can we replicate that here?

Sounds good I will update that. I did notice in Gutenberg once the post is published it shows `Switch to draft` instead:
<img width="467" alt="Screen Shot 2022-08-10 at 1 15 31 PM" src="https://user-images.githubusercontent.com/2240960/183960868-efceec13-ee38-4abb-9e6d-1dde9bc802cd.png">

Should I follow that same format instead of hiding the `Save draft` button once it has been published?

louwie17 committed:
Moved pending action variables to product helper and updated buttons to reflect new suggestions

jarekmorawski commented:
> Should I follow that same format instead of hiding the Save draft button once it has been published?

Yes! We want to be consistent with Gutenberg in this regard.

louwie17 committed:
Fix backwards compabitibility issue with Form changes

louwie17 committed:
Add switch to draft button

octaedro reviewed (approved):
Thank you for the changes! LGTM 🚀
